llama_index/llms/vllm.py

```diff
@@ -0,0 +1,374 @@
+        top_p: float = 1.0,
+        top_k: int = -1,
+        use_beam_search: bool = False,
+        stop: Optional[List[str]] = None,
+        ignore_eos: bool = False,
+        max_new_tokens: int = 512,
+        logprobs: Optional[int] = None,
+        dtype: str = "auto",
+        download_dir: Optional[str] = None,
+        vllm_kwargs: Dict[str, Any] = {},
+        callback_manager: Optional[CallbackManager] = None,
+    ) -> None:
+        self._client = None
+
+        callback_manager = callback_manager or CallbackManager([])
+
+        model = ""
+        super().__init__(
+            model=model,
+            temperature=temperature,
+            n=n,
+            best_of=best_of,
+            presence_penalty=presence_penalty,
+            frequency_penalty=frequency_penalty,
+            top_p=top_p,
+            top_k=top_k,
+            use_beam_search=use_beam_search,
+            stop=stop,
+            ignore_eos=ignore_eos,
+            max_new_tokens=max_new_tokens,
+            logprobs=logprobs,
+            dtype=dtype,
+            download_dir=download_dir,
+            vllm_kwargs=vllm_kwargs,
+            api_url=api_url,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "VllmServer"
+
+    @llm_completion_callback()
+    def complete(self, prompt: str, **kwargs: Any) -> List[CompletionResponse]:
+        kwargs = kwargs if kwargs else {}
+        params = {**self._model_kwargs, **kwargs}
+
+        from vllm import SamplingParams
+
+        responses = []
+        # build sampling parameters
+        sampling_params = SamplingParams(**params).__dict__
+        sampling_params["prompt"] = prompt
+        response = post_http_request(self.api_url, sampling_params, stream=False)
+        outputs = get_response(response)
+        for output in outputs:
+            responses.append(CompletionResponse(text=output))
+
+        return responses
+
+    @llm_completion_callback()
+    def stream_complete(self, prompt: str, **kwargs: Any) -> CompletionResponseGen:
+        kwargs = kwargs if kwargs else {}
+        params = {**self._model_kwargs, **kwargs}
+
+        from vllm import SamplingParams
+
+        # build sampling parameters
+        sampling_params = SamplingParams(**params).__dict__
+        sampling_params["prompt"] = prompt
+        response = post_http_request(self.api_url, sampling_params, stream=True)
+
+        def gen() -> CompletionResponseGen:
+            for chunk in response.iter_lines(
+                chunk_size=8192, decode_unicode=False, delimiter=b"\0"
+            ):
+                if chunk:
+                    data = json.loads(chunk.decode("utf-8"))
+
+                    yield CompletionResponse(text=data["text"][0])
+
+        return gen()
+
+    @llm_completion_callback()
+    async def acomplete(self, prompt: str, **kwargs: Any) -> CompletionResponse:
+        kwargs = kwargs if kwargs else {}
+        return self.complete(prompt, **kwargs)
```

Comment:
should you add `await` here

Reply:
I think we probably could have added a proper `async def post_http_request()` to have proper async

Right now, this is "fake" async

Can address in another PR though